tests/integration-tests/tests/cloudwatch_logging/test_cloudwatch_logging.py

```diff
@@ -0,0 +1,461 @@
+                        "is_empty": false,
+                        "tail": "2019-11-01 22:00:14.7647 TORQUE authd daemon started"
+                    }
+                },
+                "agent_status": "running"
+            }
+        }
+    }
+    """
+
+    def __init__(self, scheduler, os, cluster, data_retriever):
+        """Get the state of the cluster as it pertains to the CloudWatch logging feature."""
+        self.scheduler = scheduler
+        self.platform = self._base_os_to_platform(os)
+        self.cluster = cluster
+        self.remote_command_executor = RemoteCommandExecutor(self.cluster)
+        self.scheduler_commands = get_scheduler_commands(self.scheduler, self.remote_command_executor)
+        self.data_retriever = data_retriever
+        self._relevant_logs = {"MasterServer": [], "ComputeFleet": []}
+        self._cluster_log_state = {"MasterServer": {}, "ComputeFleet": {}}
+        self._set_cluster_log_state()
+
+    def get_logs_state(self):
+        """Get the state of the log files applicable to each of the cluster's EC2 instances."""
+        desired_keys = ["hostname", "instance_id", "node_role", "agent_status", "logs"]
+        states = [{key: self._cluster_log_state["MasterServer"][key] for key in desired_keys}]
+        states.extend([
+            {key: host_dict[key] for key in desired_keys}
+            for hostname, host_dict in self._cluster_log_state["ComputeFleet"].items()
+        ])
+        assert_that(states).is_length(self.scheduler_commands.compute_nodes_count() + 1)  # computes + master
+        return states
+
+    def _dump_cluster_log_state(self):
+        """Dump the JSON-ified string of self._cluster_log_state for debugging purposes."""
+        return json.dumps(self._cluster_log_state, indent=4)
+
+    @staticmethod
+    def _base_os_to_platform(base_os):
+        """Turn the name of a base OS into the platform."""
+        # Special case: alinux is how the config file refers to amazon linux, but in the chef cookbook
+        # (and the cloudwatch log config produced by it) the platform is "amazon".
+        translations = {'alinux': 'amazon'}
+        no_digits = base_os.rstrip(string.digits)
+        return translations.get(no_digits, no_digits)
+
+    def _set_master_instance(self, instance):
+        """Set the master instance field in self.cluster_log_state."""
+        self._cluster_log_state["MasterServer"].update({
+            "node_role": "MasterServer",
+            "hostname": instance["PrivateDnsName"],
+            "instance_id": instance["InstanceId"]
+        })
+
+    def _add_compute_instance(self, instance):
+        """Update the cluster's log state by adding a compute node."""
+        self._cluster_log_state["ComputeFleet"][instance["PrivateDnsName"]] = {
+            "node_role": "ComputeFleet",
+            "hostname": instance["PrivateDnsName"],
+            "instance_id": instance["InstanceId"]
+        }
+
+    def _get_initial_cluster_log_state(self):
```

Comment:
There's a much easier way to do this - see https://github.com/aws/aws-parallelcluster/blob/develop/cli/pcluster/commands.py#L466

Reply:
For each instance I need:
* its role in the cluster (master or compute),
* its instance ID
* its hostname

I can get the first two items in that list via the function you pointed me to (using the `LogicalResourceId` and `PhysicalResourceId` keys of the dicts in the returned list). What would be the best way to then map the instance IDs to hostnames? Call `ec2.describe_instances` with an `InstanceIds` arg obtained from the returned value of the first call? That makes this a two-step process instead of one (albeit two more reasonably sized processes).

Reply:
That makes sense. You can either call `ec2.descibe_instances()` or `asg.describe_autoscaling_instances()` which both take the instance Id's as input so it's a moot point.